File: python/24-SwapNodesInPairs/main.py
from typing import Optional
import unittest

import sys
sys.path.append('../modules')
from leetcode import ListNode

class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        
        second = head.next
        if not second:
            return head
        
        rest = second.next

        second.next = head
        head.next = self.swapPairs(rest)

        return second
    
class Tests(unittest.TestCase):
    def test_example1(self):
        solution = Solution()
        s = solution.swapPairs(ListNode.makeList(1, 2, 3, 4))
        
        values = ListNode.flatten(s)
        self.assertListEqual(values, [2, 1, 4, 3]);

    def test_example2(self):
        solution = Solution()
        s = solution.swapPairs(None)
        self.assertIsNone(s)

    def test_example3(self):
        solution = Solution()
        s = solution.swapPairs(ListNode(1))
        self.assertTrue(s.val == 1)
        self.assertIsNone(s.next)

if __name__ == '__main__':
    unittest.main()

File: cpp/24-SwapNodesInPairs/main.cpp
#include "../LeetCode.h"

class Solution 
{
public:
    ListNode* swapPairs(ListNode* head) 
    {
        if(head == nullptr || head->next == nullptr) return head;

        ListNode *swapHead = nullptr;
        ListNode *tail = nullptr;

        while(head)
        {
            if(head->next == nullptr)
            {
                if(swapHead == nullptr)
                {
                    swapHead = head;
                }
                else
                {
                    tail->next = head;
                }

                head = head -> next;

            }
            else
            {
                // We've got at least 2 items
                auto newHead = head->next->next;

                auto swap = head->next;
                swap->next = head;
                swap->next->next = nullptr;

                if(swapHead == nullptr)
                {
                    swapHead = swap;
                }
                else
                {
                    tail->next = swap;
                }

                tail = swap->next;
                head = newHead;
            }
        }

        return swapHead;
    }
};

void example1()
{
    Solution s;

    auto head = makeList({1, 2, 3, 4});
    auto swap = s.swapPairs(head);
    print(swap);
}

void example2()
{
    Solution s;

    auto head = makeList({});
    auto swap = s.swapPairs(head);
    print(swap);
}

void example3()
{
    Solution s;

    auto head = makeList({1});
    auto swap = s.swapPairs(head);
    print(swap);
}

void example4()
{
    Solution s;

    auto head = makeList({1, 2, 3, 4, 5, 6, 7});
    auto swap = s.swapPairs(head);
    print(swap);
}

int main()
{
    example1();
    example2();
    example3();
    example4();
}